### PR TITLE
Add cpplint to build

### PR DESCRIPTION
**Pull request by AndrewGaspar**
## PR Summary

I always had to look up how to run cpplint, so now cpplint runs as part of CMake. The lint target is automatically added to the default build target, but can be disabled with `PARTHENON_LINT_DEFAULT`.

## PR Checklist

- [x] Code passes cpplint (Now it must!)
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted

### CODE REVIEW

File: example/main.cpp [hunk 1]
```diff
@@ -1,131 +0,0 @@
-//========================================================================================
```
Comment (Yurlungur) [inline example/main.cpp:1]: I thought we deleted this file a while ago. Why is it showing up in the diff now?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -12,9 +12,6 @@
 # the public, perform publicly and display publicly, and to permit others to do so.
 #=========================================================================================
 
-# Configure defs.hpp
-set(PROBLEM_GENERATOR "<not-implemented>") # TODO: Figure out what to put here
-
```
Comment (Yurlungur) [inline src/CMakeLists.txt:17]: I'm not sure I would delete this. I think this is still an outstanding problem we will eventually need to solve.

File: example/main.cpp (hunk 1 above)
Comment (AndrewGaspar) [reply example/main.cpp:1]: https://github.com/lanl/parthenon/blob/master/example/main.cpp

Comment (Yurlungur) [reply example/main.cpp:1]: huh ok I could have sworn we deleted this a while ago. My bad.

**Pull request merged**